### PR TITLE
Feature/development repo

### PR DESCRIPTION
**Pull request by RaymondMouthaan**
This feature adds the possibility to `release` images and there manifest list to a new docker repo `nodered/node-red-dev`, so they can be tested before releasing to `nodered/node-red`.

- Github tags with either `dev`, `test`, `edge` or `beta` get manifest listed and pushed to `nodered/node-red-dev`

- Github tags with either `beta` or none of these (in case of a final) get manifest listed and pushed to `nodered/node-red`

Which github tag to use depends on the change and goal of releasing

### CODE REVIEW

**Comment by dceejay** [comment]
Basic idea is sound - to have a separate docker hub project for anything not stable.
We were wondering if it could perhaps be branch based - in that any released tagged from a branch would get automatically named after that branch. (and be based on the Node-RED version. So for example 1.0.1-dev) - and re-use the names - overwrite releases. And also could there be a webhook to delete releases if/when a branch is deleted ?

**Comment by RaymondMouthaan** [comment]
Current PR - tags on Docker are based on Release tags in Github, this is a human action. 
- So `v1.0.0-dev` becomes -> `1.0.0-dev` and `dev` tags on Docker Hub
- So `v1.0.0-test` becomes -> `1.0.0-test` and `test` tags on Docker Hub
- So `v1.0.0-beta` becomes -> `1.0.0-beta` and `beta` tags on Docker Hub

As long as the github tags contain the words `dev`, `test` or `beta` the manifest list tags will be overridden. Image tags only get overridden when the full tags remain the same. Note the tag must be deleted in github, before creating the new (same) tag.

Image tags based on node-red version from package.json + branch name can be done. Even 'releasing' them to docker repo `nodered/node-red-dev` without human interference could be done - otherwise I don't see the benefits of deriving the tags from node-red version from package.json + branch name. In case of a beta, a new branch should be created on github containing `beta` in it's name - same for `test`.

Not sure if docker hub webhooks allow to delete certain images / manifest lists, if it does, then this should be done in the github project (not Travis CI)

**Comment by dceejay** [comment]
Yes - betas should probably still be manual, but all others could be fully automatic (imho) - unless [skip travis] of course - especially if they could overwrite existing tags to keep them up to date.

Re cleanup - hopefully that could be a github action or somesuch...

**Comment by knolleary** [comment]
@dceejay Not sure what you mean by manual vs fully automatic. We only push to dockerhub when a tag is created... and if we don't push to dockerhub then the Travis build is kinda pointless other than to sanity check the build works without wanting to test the output.

**Comment by RaymondMouthaan** [comment]
@knolleary - not completely true that building without pushing to docker hub is pointless. With travis we can test if the images are build (check modifications of the dockerfile, travis.yml, docker.sh etc), but also we could improve the current tests, now the test just spins up the builded images and checks if it did not return other then 0 (zero), but we could improve it to test if node-red has been started or even more advanced tests. 

My proposal would be:

To publish full automatically to `nodered/node-red-dev`(if `[skip travis]` is not used in the commit message - in case of changes to the readme or anything else that doesn't need image rebuild) - no more manual releasing / tagging at github. A commit on branch node-red-dev, triggers travis to start building the images and publish them to `nodered/node-red-dev`, images tags based on the node-red version + the branch name (in this case `dev`).

A second branch `nodered/node-red-beta` for node-red beta versions, same fully automated pipeline like dev but with the difference that these images get tagged with `beta`, but also are published to `nodered/node-red-dev`. Betas images/tags are only overridden in case of a new beta of node-red `core` (like [1.0.0-beta.4](https://github.com/node-red/node-red/releases/tag/1.0.0-beta.4)) - if it is not core related then the dev branch should be used.

No direct commits to master - all goes via either `nodered/node-red-dev` or `nodered/node-red-beta` - manually create a release on github using the `v` syntax to start the pipeline for releasing to `nodered/node-red`, where only `final` versions are kept.

Users of `nodered/node-red-dev` should be informed that images + manifest lists in this repo are short term and might be changing or cleaned up, aimed for testing purpose only. 

About maintaining the `nodered/node-red-dev` repo, in case of a 'beta`, I propose to delete manually each image / manifest list manually after a merge to master and release to `nodered/node-red` or delete the whole `nodered/node-red-dev` once a month / two months (depending how much it contains).

**Comment by dceejay** [comment]
I think yes... - but would / could this handle multiple branches ? so it would also build (and release) feature branches if pr accepted etc ?

**Comment by knolleary** [comment]
Development of new features should happen in feature-specific branches and then get PRd against master (or dev initially and to master when ready to release to the main stream).

So I had imagined we'd be able to push to the `node-red-dev` dockerhub from any branch that isn't master and the image would be tagged with the branch name so we know where it came from.

If we only did that from one branch, we'd only be able to develop one feature at anytime.

 I hadn't thought we would build on every commit as that feels wasteful of time and resource. I'd prefer it were done by request via tagging unless that proves too inconvenient.

So in summary...

 - tagging on master goes to `nodered/node-red` as it does today
 - tagging on any other branch goes to `nodered/node-red-dev` and the image tag includes the branch name

**Comment by dceejay** [comment]
Basically yes... I would say....
Master to  nodered/node-red  - manually - by tag - as per today
Beta to nodered/node-red-dev - manually - by tag (so we can control carefully - do beta 1 2 3 etc)
Any other branch - to nodered/node-red-dev - automatically - to include base NR version and branch name in tag eg   1.0.1-Daves-test-branch
dev branch reserved for merging any of the these other test/feature branches into for "wider" consumption - eg review etc... but also auto build and release
Adding [skip travis] keyword to any commit name will avoid travis running (is a travis default feature) - and will thus not do an auto build or release. So we can always add that to any incoming PR we want to stop initial build if we want to.

**Comment by RaymondMouthaan** [comment]
We should agree on a naming convention for feature branches, always start with 'feature-<subject>' or 'feature-<github-number-of-issue>'?

Once feature has been created and the whole set of images and manifest lists are pushed to the repo and we did some testing... We PR it to dev... The whole process starts again with tags 'dev'. Finally dev will be PR-ed to master to get it into a "final" nodered/node-red?

^ just to make sure I got it right before changing all the files

**Comment by knolleary** [comment]
I don't necessarily see a need for a naming convention - we don't have any such requirement in our other repos.

And whether a pr goes to dev or master will depend on individual circumstances. Don't want to get too wrapped up in inventing processes here. In theory the docker images ought to be relatively stable.

**Comment by dceejay** [comment]
Ok - so Ok to update this PR to make it as described ?
And then give it a go ?

**Comment by dceejay** [comment]
Actually from here - https://docs.travis-ci.com/user/pull-requests/#how-pull-requests-are-built
this looks useful to turn off...

> To only build on push events not on pull requests, disable Build on Pull Requests in your repository settings.
> To only build pull requests targeting specific branches you can use the branches: only: key, which will also restrict the branches that trigger builds.
